### PR TITLE
Add user-email header to websocket connections

### PR DESCRIPTION
**Pull request by r4vi**
It turns out that websockets weren't going through the express middleware, this
means that they didn't get the user header added to them.

This PR does the following:
 * runs the express middleware manually against the websocket connections so
 that proxyReqWs event has access to `req.session.passport.user`
 * registers a handler for 'proxyReqWs' that adds the user header

### CODE REVIEW

File: app/server.js [hunk 1]
```diff
@@ -37,5 +37,17 @@ server.on('listening', () => {
 
 server.on('upgrade', (req, socket, head) => {
   log.info('Upgrading socket connection');
-  proxy.ws(req, socket, head);
+  // run middleware stack against the websocket connection
+  // because this doesn't go through express we have to manually
+  // run the middleware we want, in this case just enough to figure out who
+  // the logged in user is
+  config.middlewareStack['cookie-parser'](req, {}, () => {
```
Comment (r4vi) [inline app/server.js]: this is pretty ugly, open to ideas / cool js tricks to flatten it

File: app/factory.js [hunk 2]
```diff
@@ -28,7 +32,7 @@ exports.create_app = (config_override) => {
   const conf = Object.assign(config, config_override);
 
   init_app(app, conf);
-  init_middleware(app, conf);
+  const middlewareStack = init_middleware(app, conf);
 
-  return app;
+  return [app, middlewareStack];
```
Comment (r4vi) [inline app/factory.js]: could potentially set this on `config` here and not return the middlewareStack

File: app/middleware/authentication.js [hunk 3]
```diff
@@ -26,6 +26,6 @@ module.exports = (app, conf, log) => {
 
   return [
     passport.initialize(),
-    passport.session(),
+    passport.session({ pauseStream: true }),
```
Comment (r4vi) [inline app/middleware/authentication.js]: might not be required... will try to remove this change

File: app/proxy.js [hunk 4]
```diff
@@ -16,7 +16,7 @@ function onError(error, req, res) {
 proxy.on('error', onError);
 tunnel.on('error', onError);
 
-proxy.on('proxyReq', (proxyReq, req) => {
+function annotateReqWithUser(proxyReq, req) {
```
Comment (r4vi) [inline app/proxy.js]: probably needs a better name

File: app/server.js (hunk 1 above)
Comment (r4vi) [reply app/server.js]: maybe promisify?

Comment (andyhd) [reply app/server.js]: You might want a separate config entry for the list of websocket middleware, then use something like https://github.com/blakeembrey/compose-middleware

Comment (r4vi) [reply app/server.js]: that looks nice

**Comment by r4vi** [comment]
we probably need to backport some of this to rstudio-auth-proxy

File: app/factory.js [hunk 5]
```diff
@@ -16,9 +16,13 @@ function init_app(app, conf) {
...
 
 function init_middleware(app, conf) {
+  const middlewareStack = {};
   conf.middleware.forEach((name) => {
-    app.use(require(`./middleware/${name}`)(app, conf, log)); // eslint-disable-line global-require
+    const middleware = require(`./middleware/${name}`)(app, conf, log); // eslint-disable-line global-require
+    middlewareStack[name] = middleware;
+    app.use(middleware);
   });
+  return middlewareStack;
```
Comment (xoen) [inline app/factory.js]: This `middlewareStack` is returned then returned again up and up and eventually it is used to set `config.middlewareStack`. Maybe I'm missing something but you have access to `config` here, couldn't we just set it here?

Comment (r4vi) [reply app/factory.js]: yeah see my comment. I'll just set it

Comment (xoen) [reply app/factory.js]: Also, I had to move between several files to follow this, why don't we do this in `config.js`? When config is initialised at start time we can just instantiate all the middleware and have it available wherever you need this list/map.

File: app/factory.js (hunk 2 above)
Comment (xoen) [reply app/factory.js]: Oh I just saw this comment, yes I think returning when you will eventually will set it in `config.middlewareStack` is a bit of a distraction. I feel like we can also get rid of the `init_middleware()` function and do that straight in `config.js` or at least move that function and the call in `config.js`.

Comment (andyhd) [reply app/factory.js]: I think _initializing_ the middleware in `config.js` is messy - the config file contains configuration, not intialization.

File: app/factory.js (hunk 5 above)
Comment (andyhd) [reply app/factory.js]: I don't think you should pass the middleware stack around - initialize a separate one for websockets instead.

**Comment by r4vi** [comment]
they have to be the same as session parser stores sessions in memory so the
app needs to share the session parser with the higher level `server`

File: app/factory.js (hunk 2 above)
Comment (r4vi) [reply app/factory.js]: agreed - not going to initalise middleware and store it in `config`

File: app/factory.js (hunk 5 above)
Comment (r4vi) [reply app/factory.js]: using `app.set` to store it. seems like a better way than passing this up https://expressjs.com/en/api.html#app.set

**Comment by r4vi** [comment]
@xoen not a full refactor but I think this is easier to follow

File: app/factory.js [hunk 6]
```diff
@@ -32,7 +33,13 @@ exports.create_app = (config_override) => {
...
-  const middlewareStack = init_middleware(app, conf);
-
-  return [app, middlewareStack];
+  const fullMiddlewareStack = init_middleware(app, conf);
+  app.set('websocketMiddleware',
+    compose([
+      fullMiddlewareStack['cookie-parser'],
+      fullMiddlewareStack.session,
+      ...fullMiddlewareStack.authentication,
+    ]));
```
Comment (andyhd) [inline app/factory.js]: Nice!

Comment (xoen) [reply app/factory.js]: Much better :)

**Pull request merged**